dcpy/builds/plan.py

```diff
@@ -179,11 +186,63 @@ def recipe_from_yaml(path: Path) -> Recipe:
         return recipe
 
 
-def plan(recipe_file: Path) -> Path:
-    logger.info("Planning recipe")
-    lock_file = recipe_file.parent / f"{recipe_file.stem}.lock.yml"
+def repeat_recipe_from_source_data_versions(
+    version: str,
+    source_data_versions: pd.DataFrame,
+    template_recipe_path: Path | None = None,
+) -> Recipe:
+    template_recipe_path_ = template_recipe_path or Path("recipe.yml")
+    recipe = recipe_from_yaml(template_recipe_path_)
+    recipe.version = version
+    if template_recipe_path is None:
+        recipe.version_type = None
+        recipe.version_strategy = None
+        recipe.base_recipe = None
+    print(source_data_versions.columns)
+    version_by_source_data_name = {
+        name: row["version"] for name, row in source_data_versions.iterrows()
+    }
+    for ds in recipe.inputs.datasets:
+        if ds.name in version_by_source_data_name:
+            ds.version = version_by_source_data_name[ds.name]
+        else:
+            raise Exception(
+                "Dataset found in template recipe not found in historical source data versions, \
+                cannot repeat build."
+            )
+
+    return recipe
+
+
+def plan(
+    recipe_file: Path | None = None, version: str | None = None, repeat: bool = False
+) -> Path:
+    lock_file = (
+        recipe_file.parent / f"{recipe_file.stem}.lock.yml"
+        if recipe_file
+        else Path("recipe.lock.yml")
+    )
+    if recipe_file and not repeat:
+        logger.info("Planning recipe")
+        recipe = plan_recipe(recipe_file, version)
+    else:
+        if version is None:
+            raise Exception("Version must be supplied if repeating a specific build")
+        product_key = publishing.PublishKey("db-pluto", version)
```

Comment:
something like `logger.info("Repeating recipe from version {version}")` would be cool, maybe after the exception isn't raied?

Reply:
Also - other unimplemented thing right now - in these cases, we need to provide product key. See hardcoded db-pluto in this line.

Already an input to the gha. Though here we need the mapping of "product name" to "s3 folder" ("pluto" to "db-pluto" etc. Recipe files include that, so maybe the cleanest way to handle this require ALL repeats to have a base recipe, just to at least get metadata like product name from as a starting point

Reply:
That'll also resolve @alexrichey 's comment on nullable template recipe for generating build_metadata